### PR TITLE
[MIRROR] [READY] I got guns

### PR DESCRIPTION
**Pull request by ghost**
Original PR: https://github.com/tgstation/tgstation/pull/41804
----------------------------
:cl: actioninja\memager
refactor: Ballistic guns have been almost entirely reworked from a backend side. This primarily focused on moving as much as possible into the base ballistic gun and turning as many other guns into essentially varedits as possible.
add: Guns can now be racked. This is the default action unless the magazine is empty.
add: Guns now can have the magazine removed by clicking on them with an open hand while they are in hand.
tweak: Suppressors are now removed by alt clicking instead of clicking with an open hand
add: Guns now have various bolt types that all function a bit differently. Open bolts cannot have a bullet chambered with no mag, locking bolts lock back after running out of ammo, etc.
add: All ballistic gun sounds are controlled by variables instead of hardcoded usages.
tweak: The l6 LMG has a reworked control scheme. alt + click to open and close the cover, open hand to remove mag, place mag in by slapping them together, default action is rack.
tweak: Functionality that used to be snowflake code such as tactical reloads has been moved to the base gun, and can be toggled by variables.
refactor: All shotguns are now properly subpaths of the shotgun type. They still work the same.
imagedel: Gun sprites have been almost entirely overhauled to use overlays instead of states. This collapsed the L6 sprite from 20 sprites to 9 sprites to give an example.
soundadd: Remixed versions of the shotgun and base gun firesound
soundadd: New rifle firesound for l6 and moist nugget, new SMG fire sound, new sniper fire sound.
soundadd: Lots of new weapon operation sounds such as racking and bolts and the like.
imageadd: New sprites for the Riot Shotgun, Combat Shotgun, c20r, Deagle, m90, double barrel shotgun and 1911 by Memager.
imageadd: All gun sprites that were pointing to the left have been flipped to point to the right.
balance: Shotguns now can be pumped faster.
balance: Bulldog can now be tac reloaded
balance: Sawn off shotguns now have an accuracy penalty and recoil
/:cl:

[GUNS](https://www.youtube.com/watch?v=og243Dom4Sw)

This is probably way too much in one PR but it's all fairly related so fuck it. Complain if it's too much and I'll split everything that's not the functionality changes off.

Sprites:
image

Should I record a video for all the sounds? It's like 10 minutes of effort man...

Shit I still need to do:
- [x] FIX THE DOUBLE BARREL (AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA)
- [x] Fix the bulldog's magazine sprites in a way that doesn't result in snowflake shitcode
- [x] Fix the moist nugget
- [x] Make tac reloads not dump the mag on the floor.
- [x] Figure out what the hell to do with cursed bolt action and arcane barrage
- [x] Decide what to do about the L6's control scheme.
- [x] Make locking bolts actually fucking lock, ebic shitcode
- [x] Make revolvers able to be spun at any time
- [x] Convert every sprite to new formats
- [x] Make sawn weapons shittier in some way.
- [x] More sounds
- [x] More sprites
- [x] Fix the various typos I have introduced.
- [x] Bugtest this shitcode
- [x] ~~reenable the laser gatling~~

### CODE REVIEW

**Comment by Manly-Dwarf** [comment]
>those new shotgun sprites

Ugh.

**Comment by Manly-Dwarf** [comment]
Okay this PR is complete garbage. Making magazines removed by alt+clicking is complete shit, who the hell wants gun racking in the game any way
Revert the racking.

**Comment by DaydreamPuppeh** [comment]
-1

**Comment by actioninja** [comment]
> Okay this PR is complete garbage. Making magazines removed by alt+clicking is complete shit, who the hell wants gun racking in the game any way

Magazines are removed by clicking with an open hand or the default self action when the magazine is empty. Suppressors are removed by alt clicking.

**Comment by yoyobatty** [comment]
**Progress of commits for this PR to make it less shit:**
Before:
Click gun with full mag, it racks
Have to rack the chamber when putting in new mag
After: 
Click gun with full mag, it removes mag
Insert mag, auto drops the bolt, chambering a round if possible.

Shit to do:
- [x]  Old sprites
- [x] Make sure our guns work

**Comment by actioninja** [comment]
Almost everything important is rolled into procs, I'll review everything that you'd need to change if you wanted it to work like that.

File: code/modules/projectiles/guns/ballistic.dm [hunk 1]
```diff
@@ -86,59 +222,98 @@
...
-		AC.bounce_away()
-		chambered = null
-		to_chat(user, "<span class='notice'>You unload the round from \the [src]'s chamber.</span>")
-		playsound(src, "gun_slide_lock", 70, 1)
-	else
-		to_chat(user, "<span class='notice'>There's no magazine in \the [src].</span>")
-	update_icon()
+			to_chat(user, "<span class='warning'>[src] is empty!</span>")
+		return
+	if(bolt_type == BOLT_TYPE_LOCKING && bolt_locked)
```
Comment (actioninja) [inline code/modules/projectiles/guns/ballistic.dm:269]: remove 298-300: not needed

**Comment by yoyobatty** [comment]
It already does, did you not see the commits.

File: code/modules/projectiles/guns/ballistic.dm [hunk 2]
```diff
@@ -3,77 +3,213 @@
...
+	bolt_locked = FALSE
+	update_icon()
+
+/obj/item/gun/ballistic/proc/insert_magazine(mob/user, obj/item/ammo_box/magazine/AM, display_message = TRUE)
+	if(user.transferItemToLoc(AM, src))
+		magazine = AM
+		if (display_message)
+			to_chat(user, "<span class='notice'>You load a new [magazine_wording] into \the [src].</span>")
+		playsound(src, load_empty_sound, load_sound_volume, load_sound_vary)
+		drop_bolt(user)//hippie edit -- rack by default so you don't have to manually fucking do it
```
Comment (actioninja) [inline code/modules/projectiles/guns/ballistic.dm:129]:
```suggestion
	if (!chambered)
		if (bolt_locked == TRUE && bolt_type == BOLT_TYPE_LOCKING_BOLT)
			drop_bolt(user)
		else
			rack(user)
```

File: code/modules/projectiles/guns/ballistic.dm (hunk 1 above)
Comment (actioninja) [reply code/modules/projectiles/guns/ballistic.dm]: actually shotguns rely on the rack part, just 289-300

File: code/modules/projectiles/guns/ballistic.dm (hunk 2 above)
Comment (yoyobatty) [reply code/modules/projectiles/guns/ballistic.dm:129]: You realize that drop_bolt checks if it's chambered and takes care of everything that you're trying to suggest, right?

File: code/modules/projectiles/guns/ballistic.dm (hunk 1 above)
Comment (yoyobatty) [reply code/modules/projectiles/guns/ballistic.dm:269]: I'm fairly certain you would want this, you can still open the bolt

File: code/modules/projectiles/guns/ballistic.dm (hunk 2 above)
Comment (actioninja) [reply code/modules/projectiles/guns/ballistic.dm:129]: Forgot that drop_bolt uses chamber_round and not process_chamber. This makes sure that the sounds work right though. There's unique sounds for rack and drop in some cases.

Comment (yoyobatty) [reply code/modules/projectiles/guns/ballistic.dm:129]: Yep, I'll do a video showcasing it just to be sure I got everything.

**Comment by actioninja** [comment]
I think I have an original sprites converted dmi, do you want the flipped directions on old sprites or just straight old sprites?

**Comment by yoyobatty** [comment]
Original would make everyone happy. I have the files myself, I'm just going to do a git checkout for the sprites.

**Comment by actioninja** [comment]
There's a lot more going on besides different sprites unless you just want to axe the whole overlay system entirely.

**Comment by yoyobatty** [comment]
Yeah I'm aware it uses overlays that would be incompatible with the old sprites, so I'm willing to just cut them out completely.

**Comment by yoyobatty** [comment]
I made a comment on the original PR highlighting how the l6_saw to_chat proc is broken because of the lack of mob argument.

**Comment by yoyobatty** [comment]
Another thing too is you forgot to do a sanity check on `can_shoot` if you try and fire the gun without a mag it produces a runtime.

**Comment by yoyobatty** [comment]
WORKS LADS!!!!!! Old sprites, old functionality! Everything has been tested to shit! This baby is ready!

**Review by yoyobatty — APPROVED** [review]
Waiting until we confirm we want these new gun sounds and all of the old sprites. Everything feels awesome and brand new but it still has the old look. I can't believe how well this turned out.

**Comment by actioninja** [comment]
> Another thing too is you forgot to do a sanity check on `can_shoot` if you try and fire the gun without a mag it produces a runtime.

I'm actually brain dead, why I thought that would work is so stupid I'm not even going to say it.
Thinking a bit more about it and how can_shoot is supposed to work, I don't know why the magazine and ammo calls are there. Doing some testing right now and it seems to work as expected with _just_ chambered being returned.

**Comment by yoyobatty** [comment]
The head coder must be thorough even with mirrored PRs.

**Pull request merged**